src/server/viper.ts

```diff
@@ -0,0 +1,345 @@
+                                },
+                            );
+                            // await connection.sendNotification(
+                            //     new rpc.NotificationType(
+                            //         'textDocument/didSave',
+                            //     ),
+                            //     {
+                            //         textDocument: { uri: viperUri },
+                            //     },
+                            // );
+                            await connection.sendNotification(
+                                new rpc.NotificationType('Verify'),
+                                {
+                                    uri: viperUri,
+                                    backend: 'silicon',
+                                    customArgs: [
+                                        '--logLevel ERROR',
+                                        `"${resolveFilePath(viperUri)}"`,
+                                    ].join(' '),
+                                    manuallyTriggered: true,
+                                },
+                            );
+                        })
+                        .catch((err) =>
+                            console.error(
+                                `Error while communicating with Viper LSP: ${err}`,
+                            ),
+                        );
+                }
+            }, verificationDebounce);
+        } else if (existsSync(viperFile)) {
+            unlinkSync(viperFile);
+        }
+        return result.diagnostics;
+    } catch (err) {
+        console.error(
+            `Error while lifting error information from Viper: ${err}`,
+        );
+        if (existsSync(viperFile)) {
+            unlinkSync(viperFile);
+        }
+        // writeFileSync(viperFile, err?.toString() || '', 'utf8');
+        return [
+            {
+                message: String(err),
+                source: 'Motoko',
+                severity: DiagnosticSeverity.Error,
+                range: {
+                    start: { line: 0, character: 0 },
+                    end: { line: 0, character: 100 }, // First line
+                },
+            },
+        ];
+    }
+}
+
+export function invalidateViper(motokoPath: string) {
+    mocViperCache.delete(motokoPath);
+}
+
+export function resolveViperMessage(diagnostic: Diagnostic): string {
+    const { message } = diagnostic;
+    if (message.startsWith('Postcondition of __init__ might not hold')) {
+        return 'Canister invariant could not be established after initializing private fields';
+    }
+    if (message.startsWith('Exhale might fail. Assertion ')) {
+        return 'Canister invariant violated by async block';
+    }
+    const match = /^Postcondition of ([a-zA-Z0-9_]+) might not hold/.exec(
+        message,
+    );
+    if (match) {
+        const [, method] = match;
+        return `Canister invariant violated by method '${method}'`;
```

Comment:
We need to make this slightly more general, as we now get the following possibility:

<img width="509" alt="Screenshot 2023-01-04 at 20 07 18" src="https://user-images.githubusercontent.com/886924/210630710-8d4f2211-2192-424f-a728-bf10e8288308.png">

Here, it's a postcondition violation, but we cannot easily distinguish between violations of "regular" postconditions vs. those representing a canister invariant check. 

```suggestion
        return `Canister specification violated by method '${method}'`;
```